Adding Bigtable Cluster.from_pb factory.

dhermes opened the pull request:
For creating a Cluster object from a protobuf cluster returned from the API.

Also adding a helper for making sure properties are set on a protobuf object.

dhermes commented on gcloud/bigtable/cluster.py:
```diff
@@ -15,9 +15,51 @@
+    :type property_name: str
+    :param property_name: The property value to check against.
+
+    :type value: object or :data:`NoneType <types.NoneType>`
```
@tseaver Note this looks like YAGNI, but this has been pulled in from the "future" and is used to make sure values are set / correct on a PB object.

dhermes commented on gcloud/bigtable/cluster.py:
```diff
@@ -15,9 +15,51 @@
+             passed with the request (if that value is not null).
+    """
+    # Make sure `property_name` is set on the response.
+    # NOTE: HasField() doesn't work in protobuf>=3.0.0a3
```
I may actually be wrong about `HasField()`. It would be nicer than (re)writing it ourselves.

tseaver replied:
The [generated docs](https://developers.google.com/protocol-buffers/docs/reference/python/google.protobuf.message.Message-class#HasField) say that `HasField` will raise `ValueError` if the message doesn't have such a field defined, and otherwise returns True IFF the field is non-empty.  OTOH, the [more modern looking version](https://developers.google.com/protocol-buffers/docs/reference/python-generated#message) say:

> Unlike in proto2, you cannot call HasField() for a singular non-message field in proto3, and the library will throw an exception if you try to do this.

dhermes replied:
Nice find! Thanks @tseaver

tseaver commented on gcloud/bigtable/cluster.py:
```diff
@@ -15,9 +15,51 @@
+                              r'(?P<cluster_id>[a-z][-a-z0-9]*)$')
+
+
+def _require_pb_property(message_pb, property_name, value):
```
Should `value` default to None, to enable the "just give me the current value" semantics, which are all we are using at the moment?

dhermes replied:
It's fine if it does, though I don't think it matters for a non-public function. Your call.

tseaver replied:
The other option would be to split the funciton in two (since there are two purposes):

```python
def _get_pb_property_value(message_pb, property_name)
    """Return a message field value."""
    all_fields = set([field.name for field in message_pb._fields])
    if property_name not in all_fields:
        raise ValueError('Message does not contain %s.' % (property_name,))
    return getattr(message_pb, property_name)

def _assert_pb_property_value(message_pb, property_name, expected_value):
    """Raise if pb field does not exist, or does not match `expected_value`."""
    if _get_pb_property_value(message_pb, property_name) != expected_value:
        raise ValueError('Message returned %s value disagreeing '
                         'with value passed in.' % (property_name,))
```

dhermes replied:
SGTM

dhermes commented:
@tseaver PTAL. I just implemented `_get_pb_property_value` (no YAGNI) and I updated the `HasField()`  note to reflect the docs you pointed to (i.e. it doesn't work in `proto3` for non-message fields).

tseaver commented:
LGTM
